tokendirectory.go

```diff
@@ -70,195 +74,157 @@ type TokenDirectory struct {
+func (t *TokenDirectory) IsRunning() bool {
+	return atomic.LoadInt32(&t.running) == 1
 }
 
-func (f *TokenDirectory) updateSources(ctx context.Context) error {
+func (t *TokenDirectory) updateSources(ctx context.Context) error {
 	wg := &sync.WaitGroup{}
-	for _, source := range f.sources {
-		for _, chainID := range source.GetChainIDs() {
-			wg.Add(1)
-			go func(chainID uint64, source Source) {
-				defer wg.Done()
-				f.updateChainSource(ctx, source, chainID)
-			}(chainID, source)
+	for _, provider := range t.providers {
+		for _, chainID := range provider.GetChainIDs() {
+			for _, source := range provider.GetSources(chainID) {
+				wg.Add(1)
+				go func(provider Provider, chainID uint64, source string) {
+					defer wg.Done()
+					t.updateProvider(ctx, provider, chainID, source)
+				}(provider, chainID, source)
+			}
 		}
 	}
 	wg.Wait()
 	return nil
 }
 
-func (f *TokenDirectory) updateChainSource(ctx context.Context, source Source, chainID uint64) {
-	f.updateMu.Lock()
-	defer f.updateMu.Unlock()
+func (t *TokenDirectory) updateProvider(ctx context.Context, provider Provider, chainID uint64, source string) {
+	t.updateMu.Lock()
+	var err error
+	defer func() {
+		t.updateMu.Unlock()
+		if t.log != nil {
+			logger := t.log.With(
+				slog.String("provider", provider.GetID()),
+				slog.Uint64("chainId", chainID),
+				slog.String("source", source),
+			)
+			if err != nil {
+				logger.Error("failed to update provider", slog.Any("err", err))
+				return
+			}
+			logger.Debug("updated provider")
+		}
+	}()
 
 	updatedContractInfo := []ContractInfo{}
-	seen := map[string]bool{}
+	seen := map[string]struct{}{}
 
-	f.contractsMu.Lock()
-	if _, ok := f.lists[chainID]; !ok {
-		f.lists[chainID] = make(map[string]*TokenList)
+	t.contractsMu.Lock()
+	if _, ok := t.lists[chainID]; !ok {
+		t.lists[chainID] = make(map[string]*TokenList)
 	}
-	if _, ok := f.contracts[chainID]; !ok {
-		f.contracts[chainID] = make(map[prototyp.Hash]ContractInfo)
+	if _, ok := t.contracts[chainID]; !ok {
+		t.contracts[chainID] = make(map[prototyp.Hash]ContractInfo)
 	}
-	f.contractsMu.Unlock()
+	t.contractsMu.Unlock()
 
-	for _, source := range source.GetURLs(chainID) {
-		tokenList, err := f.fetchTokenList(chainID, source)
-		if err != nil {
-			log.Warn().Err(err).Msgf("failed to fetch from source %q", source)
-			continue
-		}
-
-		f.lists[chainID][source] = tokenList
-
-		for i := range tokenList.Tokens {
-			contractInfo := tokenList.Tokens[i]
+	tokenList, err := provider.FetchTokenList(ctx, chainID, source)
+	if err != nil {
+		return
+	}
+	normalizeTokens(provider, tokenList)
 
-			if contractInfo.Name == "" || contractInfo.Address == "" {
-				continue
-			}
-			if contractInfo.ChainID != chainID {
-				continue
-			}
+	t.lists[chainID][provider.GetID()] = tokenList
 
-			if contractInfo.Type == "" {
-				contractInfo.Type = strings.ToUpper(tokenList.TokenStandard)
-			}
+	for i := range tokenList.Tokens {
+		contractInfo := tokenList.Tokens[i]
```

Comment:
This is maybe more straightforward? Since `i` is used only to get token from tokenList.
```suggestion
	for _, token := range tokenList.Tokens {
```